wdl2cwl/main_oop_2.py

```diff
@@ -0,0 +1,250 @@
+    #     pass
+
+    def load_wdl_task(self, obj: WDL.Task):
+        """Load task and convert to CWL."""
+
+        cwl_inputs = self.get_cwl_inputs(obj.inputs)
+        cwl_outputs = self.get_cwl_outputs(obj.outputs)
+        docker_requirement = self.get_cwl_docker_requirements(obj.runtime["docker"])
+        test_part_11 = obj.command.parts[11]
+        test_part_18 = obj.command.parts[18]
+        test_part_20 = obj.command.parts[20]
+        test_part_22 = obj.command.parts[22]
+        test_placeholder_with_args = obj.command.parts[3]
+        cwl_command_str = self.get_cwl_command_requirements(obj.command.parts)
+        base_command = ["bash", "example.sh"]
+
+        cat_tool = cwl.CommandLineTool(
+            id=obj.name,
+            inputs=cwl_inputs,
+            requirements=docker_requirement + cwl_command_str,
+            outputs=cwl_outputs,
+            cwlVersion="v1.2",
+            baseCommand=base_command,
+        )
+
+        yaml = YAML()
+        yaml.default_flow_style = False
+        yaml.indent = 4
+        yaml.block_seq_indent = 2
+        result_stream = StringIO()
+        cwl_result = cat_tool.save()
+        scalarstring.walk_tree(cwl_result)
+        yaml.dump(cwl_result, result_stream)
+        yaml.dump(cwl_result, sys.stdout)
+
+        return result_stream.getvalue()
+
+    def get_cwl_docker_requirements(self, wdl_docker: WDL.Expr):
+        requirements: List[cwl.ProcessRequirement] = []
+        dockerpull = wdl_docker.expr.referee.expr.literal.value
+        requirements.append(cwl.DockerRequirement(dockerPull=dockerpull))
+        return requirements
+
+    def get_cwl_command_requirements(self, wdl_commands: List[str]) -> List[cwl.InitialWorkDirRequirement]:
+        """Translate WDL commands into CWL Initial WorkDir REquirement."""
+        command_str: str = ""
+        for wdl_command in wdl_commands:
+            if isinstance(wdl_command, str):
+                command_str += self.translate_wdl_str(wdl_command)
+            elif isinstance(wdl_command, WDL.Expr.Placeholder):
+                command_str += self.translate_wdl_placeholder(wdl_command)
+        return [
+            cwl.InitialWorkDirRequirement(
+                listing=[cwl.Dirent(entry=command_str, entryname="example.sh")]
+            )
+        ]
+    def translate_wdl_placeholder(self, wdl_placeholder) -> str:
+        cwl_command_str = ""
+        if isinstance(wdl_placeholder.expr, WDL.Expr.Get):
+            placeholder_name = wdl_placeholder.expr.expr.name
+            if wdl_placeholder.options:
+                if "true" in wdl_placeholder.options:
+                    cwl_command_str = (
+            "$(inputs."
+            + placeholder_name
+            + " ? "
+            + '"{true_value}" : "{false_value}")'.format(
+                true_value=wdl_placeholder.options["true"],
+                false_value=wdl_placeholder.options["false"],
+            )
+            + ")"
+        )
```

Comment:
@Th3nn3ss in today's meeting you mentioned this placeholder function was getting too long. It's really (really really really!) good that you are thinking about that while you write your code. It helps maintaining the code, and is extremely helpful to other developers when the code is simpler.

Maybe we could introduce f-strings here? I had a quick look and couldn't see a lot of f-strings in this file, so perhaps that's something that you would be interested in adopting too.

Here's how I think the code above could be refactored (needs testing):

```suggestion
                if "true" in wdl_placeholder.options:
                    true_value = wdl_placeholder.options["true"]
                    false_value = wdl_placeholder.options["false"]
                    cwl_command_str = f'$(inputs.{placeholder_name} ? "{true_value}" : "{false_value}")'
```

Reply:
More about f-strings in Python:

- <https://realpython.com/python-f-strings/>
- <https://www.python.org/dev/peps/pep-0498/> (PEP-498 added in 3.6)

Reply:
That's very true. I should probably be more consistent with f-strings as it's more readable.